lib/Zonemaster/Backend/RPCAPI.pm

```diff
@@ -193,10 +200,10 @@ sub get_data_from_parent_zone {
 
 
 sub _check_domain {
-    my ( $self, $domain, $type ) = @_;
+    my ( $self, $domain ) = @_;
 
     if ( !defined( $domain ) ) {
-        return ( $domain, { status => 'nok', message => encode_entities( "$type required" ) } );
+        return ( $domain, { status => 'nok', message => N__ 'Domain name required' } );
```

Comment:
Does this `N__` translation function use MO files, compiled from PO files, in the same way as in Engine? In Engine (and CLI) it is possible to extract the strings to be translated. Does it work the same way here? Could you provide an example file with translations into French?

Reply:
`N__` is a no op function to allow extraction of string for translation ([realted doc](https://metacpan.org/pod/Locale::TextDomain#N__(ARG1))). The real translation is done [line 737](https://github.com/zonemaster/zonemaster-backend/pull/795/files/a2c13b9199165c9885ebbe70ddff3e7e3523787f#diff-f159417cb7e91c84742d1cb44858b1eed39152de169ecc78fa6677fe879ac1b2R737). 

I will provide a base translation in French once I have a better setup regarding i18n (ie add makefile for extraction and generation, and fixing paths)

Reply:
Method `start_domain_test` does not include the language that the message is to be translated to. Do you suggest that to be added?

One option is to let Backend return codes like the messages tags, e.g. "DOMAIN_NON_ASCII_NO_IDN_SUPPORT", that should be fixed and not updated, and then use the normal translation mechanism in GUI (adding strings to en.json, fr.json, sv.json etc). This means that it is up to the client to produce the message. If we add new error messages (error codes) the client has to be updated.

The other option is to include language code in the calls and let Backend translate. But what should happen then if the user changes language? When the result is presented Backend re-requests for the result with a new language tag.

Today we have to solutions for translation files, one is the PO/MO files used by Engine and CLI, the other are the language specific json files used by GUI. It would be best of we could have a solution that is the same as one of the two. That could be json files in Backend if that is the best solution.

Reply:
My current implementation uses PO/MO files on the backend. I personally prefer server-side translation because it allows for internationalization of multiple client (eg. a cli like zmtest / zmb could display the correct error message out of the box, without the need of duplicated translation files). It would probably use a parameter in the request, or a more ideal solution IMO is to use the accept-language header, which doesn't need to make any change in the request (but as we are already using a lang param for the result, we could stick to that solution to not have multiple way of doing similar stuff, I guess).

The only remaining point being if someone change their language after submitting a erroneous form. I haven't found a good solution yet. I think resending the form on the user behalf is not good UX, especially if they are in the process of correcting it.

On the other hand, having a unified translation file for client-side and server-side error is not a bad idea to have unified error messages without having to maintain duplicates translations. In that regard, I find your idea of having the backend serving json translation files interesting as it allows for centralized translation between clients without too much duplicates, beside the translation logic it self, and also allows for translation/validation updates without updating the GUI.

Reply:
There are pros and cons on putting the messages in Backend and GUI, but if we use PO/MO on Backend that should work fine, but we have to add the infrastructure there too to make it work smothly. The solution from Engine should be copied.

We should have the same solution to transfer selected language in all methods that need language. If we change to accept-language header it will mean breaking changes.

I do not think that not being able to change language is a show-stopper, so we should evaluate pros and cons.

Reply:
I have a semi-working implementation of translation with PO/MO files in the backend, heavily inspired by what has been done in the engine. Before going to far I want to test the client side translation with backend served json files, to see if we can benefit from that.